src/fairchem/core/datasets/base_dataset.py

```diff
@@ -0,0 +1,227 @@
+    Any,
+    NamedTuple,
+    TypeVar,
+)
+
+import numpy as np
+import torch
+from torch import randperm
+from torch.utils.data import Dataset
+from torch.utils.data import Subset as Subset_
+
+from fairchem.core.common.registry import registry
+
+if TYPE_CHECKING:
+    from collections.abc import Sequence
+
+    from numpy.typing import ArrayLike
+
+
+T_co = TypeVar("T_co", covariant=True)
+
+
+class DatasetMetadata(NamedTuple):
+    natoms: ArrayLike | None = None
+
+
+class UnsupportedDatasetError(ValueError):
+    pass
+
+
+class BaseDataset(Dataset[T_co], metaclass=ABCMeta):
+    """Base Dataset class for all OCP datasets."""
+
+    def __init__(self, config: dict):
+        """Initialize
+
+        Args:
+            config (dict): dataset configuration
+        """
+        self.config = config
+        self.paths = []
+
+        if "src" in self.config:
+            if isinstance(config["src"], str):
+                self.paths = [Path(self.config["src"])]
+            else:
+                self.paths = tuple(Path(path) for path in config["src"])
+
+        self.lin_ref = None
+        if self.config.get("lin_ref", False):
+            lin_ref = torch.tensor(
+                np.load(self.config["lin_ref"], allow_pickle=True)["coeff"]
+            )
+            self.lin_ref = torch.nn.Parameter(lin_ref, requires_grad=False)
+
+    def __len__(self) -> int:
+        return self.num_samples
+
+    def metadata_hasattr(self, attr) -> bool:
+        if self._metadata is None:
+            return False
+        return hasattr(self._metadata, attr)
+
+    @cached_property
+    def indices(self):
+        return np.arange(self.num_samples, dtype=int)
+
+    @cached_property
+    def _metadata(self) -> DatasetMetadata:
+        # logic to read metadata file here
+        metadata_npzs = []
+        if self.config.get("metadata_path", None) is not None:
+            metadata_npzs.append(
+                np.load(self.config["metadata_path"], allow_pickle=True)
+            )
+
+        else:
+            for path in self.paths:
+                if path.is_file():
+                    metadata_file = path.parent / "metadata.npz"
+                else:
+                    metadata_file = path / "metadata.npz"
+                if metadata_file.is_file():
+                    metadata_npzs.append(np.load(metadata_file, allow_pickle=True))
+
+        if len(metadata_npzs) == 0:
+            logging.warning(
+                f"Could not find dataset metadata.npz files in '{self.paths}'"
+            )
+            return None
+
+        metadata = DatasetMetadata(
+            **{
+                field: np.concatenate([metadata[field] for metadata in metadata_npzs])
+                for field in DatasetMetadata._fields
+            }
+        )
+
+        assert metadata.natoms.shape[0] == len(
+            self
+        ), "Loaded metadata and dataset size mismatch."
+
+        return metadata
+
+    def get_metadata(self, attr, idx):
```

Comment:
This breaks `LmdbDataset` because `get_metadata` is also defined there - https://github.com/FAIR-Chem/fairchem/blob/c73a88510ea819f56092d6145edaac27fc4ca0a4/src/fairchem/core/datasets/lmdb_dataset.py#L173. I think support for that dataset class is important as well given how common it's used.

Reply:
This totally flew under the radar and tests. Earlier we would access metadata field directly and sometimes use get_metadata, however get_metadata means two very different things for example in basedataset, 

https://github.com/FAIR-Chem/fairchem/blob/fm-v4/src/fairchem/core/datasets/base_dataset.py#L139

and then in asedataset
https://github.com/FAIR-Chem/fairchem/blob/fm-v4/src/fairchem/core/datasets/ase_datasets.py#L175

I talked to Zach and renamed the one in asedataset as sample_property_metadata , I did the same with the one in lmdb_dataset . Now lmdb_dataset inherits the correct get_metadata. 

I will add tests for this to make sure we dont miss breaking changes going forward